server: added protobuf descriptor validate support

ramaraochavali opened the pull request:
Signed-off-by: Rama <rama.rao@salesforce.com>

*Description*: Fixes https://github.com/envoyproxy/envoy/issues/2528
*Risk Level*: Low 
*Testing*: Manual
*Docs Changes*: NA
*Release Notes*: NA

ramaraochavali commented on test/common/config/grpc_mux_impl_test.cc:
```diff
@@ -30,6 +30,8 @@ namespace {
 public:
   GrpcMuxImplTest() : async_client_(new Grpc::MockAsyncClient()), timer_(new Event::MockTimer()) {
+    // Hack to force linking of the service: https://github.com/google/protobuf/issues/4221
+    envoy::service::discovery::v2::AdsDummy dummy;
```
should I include `source/server/configuration_impl.h` here and in `descriptor_test` so that this hack is not sprayed across multiple places? or is there a common include where we can have this hack?

htuch replied:
Yeah, as Matt says, let's just move this into the main binary, so we no longer need it there. I'm hopeful that we can figure out some `alwayslink` based solution, but for now this is fine. Thanks for taking on this technical debt!

mattklein123 replied:
Tests are in different binaries with different linkages, so that is expected.

mattklein123 commented:
@ramaraochavali thanks for working on this. Here is what I would do, if you don't mind.

1. Just fully delete https://github.com/envoyproxy/envoy/blob/master/test/common/protobuf/descriptor_test.cc
2. Create a new file "proto_descriptors.cc" in https://github.com/envoyproxy/envoy/tree/master/source/server. 
3. Copy all the code in my test into a function called something like `validateProtoDescriptors()` in that file.
4. Make that function return true if all the descriptors can be found.
5. Do `RELEASE_ASSERT(validateProtoDescriptors());` from https://github.com/envoyproxy/envoy/blob/master/source/exe/main_common.cc.

Thank you!

ramaraochavali commented:
@mattklein123 modified the way you suggested. PTAL.

mattklein123 commented on source/exe/main_common.cc:
```diff
@@ -42,6 +42,8 @@ class ProdComponentFactory : public ComponentFactory {
 
 } // namespace Server
 
+bool validateProtoDescriptors();
```
nit: Please add a header file and include it.

mattklein123 commented on source/server/proto_descriptors.cc:
```diff
@@ -42,8 +41,10 @@ TEST(ProtoDescriptorTest, BackCompat) {
-    EXPECT_NE(Protobuf::DescriptorPool::generated_pool()->FindMethodByName(method), nullptr)
-        << fmt::format("{} not found", method);
+    if (Protobuf::DescriptorPool::generated_pool()->FindMethodByName(method) == nullptr) {
+      descriptors_valid = false;
```
nit: just return false immediately. You can delete the local variable, and then just return true at the end of the function.

ramaraochavali commented:
@mattklein123 done. should be good now. PTAL.

mattklein123 commented on source/server/proto_descriptors.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
```
super nit: newline after this line

ramaraochavali replied:
fixed

mattklein123 commented on source/server/proto_descriptors.cc:
```diff
@@ -11,14 +11,12 @@
 #include "common/common/fmt.h"
```
Please include proto_descriptors.h here.

ramaraochavali replied:
done

mattklein123 commented on source/server/proto_descriptors.h:
```diff
@@ -0,0 +1,4 @@
+#pragma once
+namespace Envoy {
```
Sorry I didn't catch this before, please also put this inside the `Server` namespace. (Envoy::Server::validate...)

ramaraochavali replied:
no probs. added now

htuch commented:
@ramaraochavali if you want to factor out the dummy hacks, put them somewhere like `source/common/config/protobuf_link_hacks.h` so we clearly flag the inclusion as being motivated by this, that would be the most maintainable. Thanks!

ramaraochavali commented:
@htuch yes. I was not completely happy about spraying this hack across and that is what I was asking originally to have this hack in some common include. I like this idea. I have added that file and removed the hack from all other places. PTAL.

htuch commented on source/common/config/protobuf_link_hacks.h:
```diff
@@ -0,0 +1,12 @@
+
+// Hack to force linking of the service: https://github.com/google/protobuf/issues/4221.
+// This file should be included ONLY if this hack is required.
+envoy::service::discovery::v2::AdsDummy dummy;
```
Nit: call this `_als_dummy` or something to avoid over shadowing other dummy vars in files that this gets included into (same scheme for `rls_dummy`).

htuch replied:
Also make this `const` for sanity.

htuch commented on source/server/proto_descriptors.h:
```diff
@@ -0,0 +1,7 @@
+
+namespace Envoy {
+namespace Server {
+bool validateProtoDescriptors();
```
Can you add a brief comment here on what this does? Maybe just move the comment from the function definition.

ramaraochavali commented:
@htuch addressed the minor comments. added the ifdef to avoid duplicate inclusions for the `protobuf_link_hacks.h` (otherwise it failed in coverage build with multiple def error across `grpc_mux_impl_test` and `ads_integration_test`)

htuch commented on source/common/config/protobuf_link_hacks.h:
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#ifndef PROTOBUF_LINKS_
+#define PROTOBUF_LINKS_
```
I see, for coverage this is broken as they are both in the global namespace. What if you use these protos in local vars in an `inline` method that is defined here, then instantiate the function in the various modules? That should avoid introducing any globals?

I'm not sure how PROTOBUF_LINKS_ define works here, since #pragma once is a similar thing.

ramaraochavali commented:
@htuch looks like `const` did the trick. Sorry I tried both `const` and `ifdef` together. Now all tests are passing with out `ifdef`. So we are good to go.
